# 0 Installation

Please install the torch with specific CUDA version by following the instruction in the official website: https://pytorch.org/get-started/locally/

![install](./src/install.png)

In [1]:
import torch
import os

In [2]:
print(f"PyTorch version: {torch.__version__}")
print(f"Is CUDA available: {torch.cuda.is_available()}")
print(f"CUDA version: {torch.version.cuda}")
print(f"Device count: {torch.cuda.device_count()}")
print(f"Current device: {torch.cuda.current_device()}")

PyTorch version: 2.4.1
Is CUDA available: True
CUDA version: 12.1
Device count: 2
Current device: 0


In [3]:
os.environ['TORCH'] = torch.__version__
print(torch.__version__)

2.4.1


# Install PyTorch Geometric

https://pytorch-geometric.readthedocs.io/en/latest/install/installation.html

In [4]:
%pip install torch_geometric

Note: you may need to restart the kernel to use updated packages.


## Additional Libraries

If you want to utilize the full set of features from PyG, there exists several additional libraries you may want to install:

- [pyg-lib](https://github.com/pyg-team/pyg-lib): Heterogeneous GNN operators and graph sampling routines
- [torch-scatter](https://github.com/rusty1s/pytorch_scatter): Accelerated and efficient sparse reductions
- [torch-sparse](https://github.com/rusty1s/pytorch_sparse): SparseTensor support, see here
- [torch-cluster](https://github.com/rusty1s/pytorch_cluster): Graph clustering routines
- [torch-spline-conv](https://github.com/rusty1s/pytorch_spline_conv): SplineConv support

These packages come with their own CPU and GPU kernel implementations based on the  [PyTorch C++/CUDA/hip(ROCm) extension interface](https://github.com/pytorch/extension-cpp/). For a basic usage of PyG, these dependencies are fully optional. We recommend to start with a minimal installation, and install additional dependencies once you start to actually need them.

## Installation from Wheels

For ease of installation of these extensions, we provide pip wheels for these packages for all major OS, PyTorch and CUDA combinations, see [here](https://data.pyg.org/whl/):

In [5]:
import torch

print(torch.__version__)    # check torch version
print(torch.version.cuda)   # check cuda version

2.4.1
12.1


Then install the packages via pip:

```bash
pip install pyg_lib torch_scatter torch_sparse torch_cluster torch_spline_conv -f https://data.pyg.org/whl/torch-${TORCH}+${CUDA}.html
```

![pyg](./src/pyg.png)

since our
- torch.__version__: 2.4.1 
- torch.version.cuda: 12.1

We need to replace `${TORCH}` with `2.4.0` and `${CUDA}` with `cu121` in the above command.
    
```bash
pip install pyg_lib torch_scatter torch_sparse torch_cluster torch_spline_conv -f https://data.pyg.org/whl/torch-2.4.0+cu121.html
```

In [7]:
%pip install pyg_lib torch_scatter torch_sparse torch_cluster torch_spline_conv -f https://data.pyg.org/whl/torch-2.4.0+cu121.html

Looking in links: https://data.pyg.org/whl/torch-2.4.0+cu121.html
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.5/2.5 MB 5.5 MB/s eta 0:00:00a 0:00:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 10.9/10.9 MB 14.4 MB/s eta 0:00:00 0:00:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.1/5.1 MB 8.8 MB/s eta 0:00:00a 0:00:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.4/3.4 MB 4.4 MB/s eta 0:00:00a 0:00:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 989.6/989.6 kB 3.3 MB/s eta 0:00:00a 0:00:01
  Using cached scipy-1.14.1-cp312-cp312-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (60 kB)
Using cached scipy-1.14.1-cp312-cp312-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (40.8 MB)
Note: you may need to restart the kernel to use updated packages.
